File: src/data/couleurs.ts
// =============================================================================
//  Couleurs proposées dans les formulaires d'annonce.
//
//  Une liste FERMÉE, et courte à dessein : « bleu alpin », « graphite sidéral »
//  et autres noms marketing vont dans le titre ou la description. Ici, il
//  s'agit de cocher vite et de filtrer juste — quinze teintes suffisent à
//  décrire ce qui se vend sur le marché.
//
//  `css` sert à peindre la pastille (formulaire ET fiche) : montrer la couleur
//  vaut mieux que l'écrire. Les teintes claires portent `clair` pour recevoir
//  un liseré, sans quoi « Blanc » disparaît sur fond blanc.
// =============================================================================

export interface Couleur {
  nom: string
  css: string
  /** Teinte claire : la pastille reçoit un liseré pour rester visible. */
  clair?: boolean
}

export const COULEURS: Couleur[] = [
  { nom: 'Noir', css: '#1B1A17' },
  { nom: 'Blanc', css: '#FFFFFF', clair: true },
  { nom: 'Gris', css: '#9CA3AF' },
  { nom: 'Argenté', css: 'linear-gradient(135deg, #F3F4F6, #9CA3AF)', clair: true },
  { nom: 'Doré', css: 'linear-gradient(135deg, #F5D882, #D4AF37)', clair: true },
  { nom: 'Bleu', css: '#2563EB' },
  { nom: 'Vert', css: '#16A34A' },
  { nom: 'Rouge', css: '#DC2626' },
  { nom: 'Orange', css: '#F97316' },
  { nom: 'Jaune', css: '#FACC15', clair: true },
  { nom: 'Rose', css: '#EC4899' },
  { nom: 'Violet', css: '#8B5CF6' },
  { nom: 'Marron', css: '#92400E' },
  { nom: 'Beige', css: '#E8D8C3', clair: true },
  { nom: 'Multicolore', css: 'conic-gradient(#F77F00, #FACC15, #16A34A, #2563EB, #8B5CF6, #EC4899, #F77F00)' },
]

/**
 * Le répertoire des couleurs connues — celui qui permet de relire une annonce.
 *
 * Les quinze teintes ci-dessus ne suffisent plus : chaque sous-catégorie
 * propose la palette de son métier. Un meuble se décline en iroko, teck et
 * wengé ; un fond de teint en carnations ; une mèche en numéros (1B, 27, 613).
 * Une annonce enregistrée avec « Iroko » doit pouvoir se relire des mois plus
 * tard — sur la fiche, dans les filtres — sans qu'on sache de quelle
 * sous-catégorie elle vient. Il faut donc les connaître toutes.
 */
const PAR_NOM: Record<string, Couleur> = Object.fromEntries(COULEURS.map((c) => [c.nom, c]))

/**
 * Ajoute les palettes des métiers au répertoire.
 *
 * Pourquoi un enregistrement, et non un simple `import` : ce fichier ne doit
 * RIEN emprunter à `data/sous/`, qui lui emprunte déjà `COULEURS`. Les deux
 * s'important l'un l'autre, celui qui démarre le second trouve le premier à
 * moitié construit — `COULEURS` vaut `undefined` au moment où on le parcourt,
 * et l'application ne s'ouvre plus. Le sens de lecture est donc à sens
 * unique : `sous/` connaît les couleurs, les couleurs ne connaissent pas
 * `sous/` et attendent qu'on vienne les compléter.
 *
 * Un nom déjà connu n'est pas remplacé : les quinze teintes générales font foi.
 */
export function enregistrerCouleurs(liste: Couleur[]): void {
  for (const c of liste) if (!PAR_NOM[c.nom]) PAR_NOM[c.nom] = c
}

/** La couleur portant ce nom, ou null pour une valeur libre / inconnue. */
export function couleurParNom(nom: string): Couleur | null {
  return PAR_NOM[nom.trim()] ?? null
}

/** « Noir, Bleu » → [Couleur, Couleur] — les noms inconnus sont ignorés sans bruit. */
export function lireCouleurs(v?: string): Couleur[] {
  return (v ?? '')
    .split(',')
    .map((s) => couleurParNom(s))
    .filter((c): c is Couleur => c !== null)
}

// -----------------------------------------------------------------------------
//  Variantes par couleur.
//
//  Une couleur cochée peut porter SA photo, SON prix et SES détails — le
//  vendeur qui a le même téléphone en noir et en bleu ne publie qu'une
//  annonce, mais l'acheteur voit chaque version.
//
//  Le stockage réutilise les attributs texte de l'annonce, sur le modèle des
//  numéros de documents fonciers (`num_<doc>`) :
//    var_Noir_photo = "2"        → index dans les photos de l'annonce
//    var_Noir_prix  = "85000"    → prix propre à cette couleur (sinon le prix principal)
//    var_Noir_note  = "128 Go"   → détails propres à cette couleur
//  Rien à changer côté serveur : ce sont des paires clé/valeur comme les autres.
// -----------------------------------------------------------------------------

export type ChampVariante = 'photo' | 'prix' | 'note'

export function cleVariante(nom: string, champ: ChampVariante): string {
  return `var_${nom}_${champ}`
}

export interface Variante {
  couleur: Couleur
  /** Index de la photo dans `images`, ou null si aucune photo n'est liée. */
  photo: number | null
  prix: number | null
  note: string
}

/** Les variantes déclarées, dans l'ordre des couleurs cochées. */
export function lireVariantes(attrs: Record<string, string>, nbPhotos: number): Variante[] {
  return lireCouleurs(attrs.couleurs).map((c) => {
    const rawPhoto = attrs[cleVariante(c.nom, 'photo')] ?? ''
    const photo = /^\d+$/.test(rawPhoto) && Number(rawPhoto) < nbPhotos ? Number(rawPhoto) : null
    const rawPrix = (attrs[cleVariante(c.nom, 'prix')] ?? '').replace(/\D/g, '')
    return {
      couleur: c,
      photo,
      prix: rawPrix !== '' ? Number(rawPrix) : null,
      note: (attrs[cleVariante(c.nom, 'note')] ?? '').trim(),
    }
  })
}

/** Au moins une couleur porte une photo, un prix ou une note. */
export function aDesVariantes(attrs: Record<string, string>, nbPhotos: number): boolean {
  return lireVariantes(attrs, nbPhotos).some((v) => v.photo !== null || v.prix !== null || v.note !== '')
}

File: flutter_app/lib/data/formulaires/couleurs.dart
// =============================================================================
//  Les couleurs des formulaires — port fidèle de src/data/couleurs.ts + les
//  palettes de métier de src/data/sous/mode.ts.
//
//  Une couleur cochée peut porter SA photo, SON prix, SES détails et SES tailles
//  restantes : le vendeur qui a le même article en noir et en bleu ne publie
//  qu'une annonce, mais l'acheteur voit chaque version. Le stockage réutilise
//  les attributs texte, sur le modèle des numéros de documents fonciers :
//      var_Noir_photo = "2"      → index dans les photos de l'annonce
//      var_Noir_prix  = "85000"  → prix propre à cette couleur
//      var_Noir_note  = "128 Go" → détails propres à cette couleur
//      var_Noir_tailles = "40, 42" → un champ déclinable (les tailles) par coloris
//  Rien à changer côté serveur : ce sont des paires clé/valeur comme les autres.
// =============================================================================
import 'package:flutter/material.dart';

/// Une teinte proposée dans le formulaire. La pastille se peint (`couleur` unie
/// ou `degrade`) ; `clair` reçoit un liseré pour rester visible sur fond clair.
class Couleur {
  final String nom;
  final Color? couleur;
  final Gradient? degrade;
  final bool clair;
  const Couleur(this.nom, {this.couleur, this.degrade, this.clair = false});
}

// Les dégradés récurrents, définis une fois.
const _argente = LinearGradient(
    begin: Alignment.topLeft,
    end: Alignment.bottomRight,
    colors: [Color(0xFFF3F4F6), Color(0xFF9CA3AF)]);
const _dore = LinearGradient(
    begin: Alignment.topLeft,
    end: Alignment.bottomRight,
    colors: [Color(0xFFF5D882), Color(0xFFD4AF37)]);
const _multicolore = SweepGradient(colors: [
  Color(0xFFF77F00),
  Color(0xFFFACC15),
  Color(0xFF16A34A),
  Color(0xFF2563EB),
  Color(0xFF8B5CF6),
  Color(0xFFEC4899),
  Color(0xFFF77F00),
]);

/// Les quinze teintes générales — la palette par défaut.
const List<Couleur> couleursBase = [
  Couleur('Noir', couleur: Color(0xFF1B1A17)),
  Couleur('Blanc', couleur: Color(0xFFFFFFFF), clair: true),
  Couleur('Gris', couleur: Color(0xFF9CA3AF)),
  Couleur('Argenté', degrade: _argente, clair: true),
  Couleur('Doré', degrade: _dore, clair: true),
  Couleur('Bleu', couleur: Color(0xFF2563EB)),
  Couleur('Vert', couleur: Color(0xFF16A34A)),
  Couleur('Rouge', couleur: Color(0xFFDC2626)),
  Couleur('Orange', couleur: Color(0xFFF97316)),
  Couleur('Jaune', couleur: Color(0xFFFACC15), clair: true),
  Couleur('Rose', couleur: Color(0xFFEC4899)),
  Couleur('Violet', couleur: Color(0xFF8B5CF6)),
  Couleur('Marron', couleur: Color(0xFF92400E)),
  Couleur('Beige', couleur: Color(0xFFE8D8C3), clair: true),
  Couleur('Multicolore', degrade: _multicolore),
];

/// Les numéros de mèches qu'on emploie à Adjamé et à Cocody.
const List<Couleur> paletteCheveux = [
  Couleur('1 · Noir de jais', couleur: Color(0xFF0A0A0A)),
  Couleur('1B · Noir naturel', couleur: Color(0xFF14100E)),
  Couleur('2 · Brun très foncé', couleur: Color(0xFF2A1A12)),
  Couleur('4 · Châtain', couleur: Color(0xFF5A3722)),
  Couleur('6 · Châtain clair', couleur: Color(0xFF7B4E2E)),
  Couleur('27 · Miel', couleur: Color(0xFFB87A3D)),
  Couleur('30 · Cuivré', couleur: Color(0xFFA85520)),
  Couleur('33 · Acajou', couleur: Color(0xFF6E2B22)),
  Couleur('99J · Bordeaux', couleur: Color(0xFF5A1224)),
  Couleur('613 · Blond platine', couleur: Color(0xFFE4CFA1), clair: true),
  Couleur('Gris / argent', couleur: Color(0xFFA8A29E)),
  Couleur('Ombré / mèches',
      degrade: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          colors: [Color(0xFF14100E), Color(0xFFB87A3D)],
          stops: [0.45, 1.0])),
  Couleur('Coloré (fantaisie)',
      degrade: SweepGradient(colors: [
        Color(0xFFEC4899),
        Color(0xFF8B5CF6),
        Color(0xFF2563EB),
        Color(0xFF16A34A),
        Color(0xFFFACC15),
        Color(0xFFEC4899),
      ])),
];

/// Les carnations, du plus clair au plus profond (fonds de teint, poudres).
const List<Couleur> paletteTeint = [
  Couleur('Très clair', couleur: Color(0xFFF0D5BE), clair: true),
  Couleur('Clair', couleur: Color(0xFFE8C39E), clair: true),
  Couleur('Beige doré', couleur: Color(0xFFD9A97A), clair: true),
  Couleur('Miel', couleur: Color(0xFFC68A4E)),
  Couleur('Caramel', couleur: Color(0xFFB2703A)),
  Couleur('Noisette', couleur: Color(0xFF96562C)),
  Couleur('Cacao', couleur: Color(0xFF7A4423)),
  Couleur('Chocolat', couleur: Color(0xFF5E3319)),
  Couleur('Ébène', couleur: Color(0xFF452412)),
  Couleur('Ébène profond', couleur: Color(0xFF2E1709)),
];

/// Lèvres, ongles, yeux : les noms des nuanciers (rouge à lèvres, vernis…).
const List<Couleur> paletteMaquillage = [
  Couleur('Nude', couleur: Color(0xFFD6A48C), clair: true),
  Couleur('Rose', couleur: Color(0xFFEC4899)),
  Couleur('Fuchsia', couleur: Color(0xFFC2185B)),
  Couleur('Corail', couleur: Color(0xFFFF6F52)),
  Couleur('Rouge', couleur: Color(0xFFDC2626)),
  Couleur('Bordeaux', couleur: Color(0xFF7A1F2B)),
  Couleur('Prune', couleur: Color(0xFF6B2D5B)),
  Couleur('Marron', couleur: Color(0xFF92400E)),
  Couleur('Orange', couleur: Color(0xFFF97316)),
  Couleur('Doré', degrade: _dore, clair: true),
  Couleur('Argenté', degrade: _argente, clair: true),
  Couleur('Noir', couleur: Color(0xFF1B1A17)),
  Couleur('Bleu', couleur: Color(0xFF2563EB)),
  Couleur('Vert', couleur: Color(0xFF16A34A)),
  Couleur('Violet', couleur: Color(0xFF8B5CF6)),
  Couleur('Transparent',
      degrade: LinearGradient(
          begin: Alignment.topLeft,
          end: Alignment.bottomRight,
          colors: [Color(0xFFFFFFFF), Color(0xFFE5E7EB)]),
      clair: true),
  Couleur('Pailleté',
      degrade: SweepGradient(colors: [
        Color(0xFFFFF3C4),
        Color(0xFFD4AF37),
        Color(0xFFFFFFFF),
        Color(0xFFD4AF37),
        Color(0xFFFFF3C4),
      ]),
      clair: true),
];

/// La palette de base de l'Électronique — les quinze teintes, plus le « gris
/// sidéral » (space grey) et le « transparent » qu'on voit sur les coques et
/// certains téléphones.
const List<Couleur> paletteElectronique = [
  Couleur('Noir', couleur: Color(0xFF1B1A17)),
  Couleur('Blanc', couleur: Color(0xFFFFFFFF), clair: true),
  Couleur('Gris', couleur: Color(0xFF9CA3AF)),
  Couleur('Gris sidéral', couleur: Color(0xFF4B5057)),
  Couleur('Argenté', degrade: _argente, clair: true),
  Couleur('Doré', degrade: _dore, clair: true),
  Couleur('Bleu', couleur: Color(0xFF2563EB)),
  Couleur('Vert', couleur: Color(0xFF16A34A)),
  Couleur('Rouge', couleur: Color(0xFFDC2626)),
  Couleur('Orange', couleur: Color(0xFFF97316)),
  Couleur('Jaune', couleur: Color(0xFFFACC15), clair: true),
  Couleur('Rose', couleur: Color(0xFFEC4899)),
  Couleur('Violet', couleur: Color(0xFF8B5CF6)),
  Couleur('Marron', couleur: Color(0xFF92400E)),
  Couleur('Beige', couleur: Color(0xFFE8D8C3), clair: true),
  Couleur('Transparent',
      degrade: LinearGradient(
          begin: Alignment.topLeft,
          end: Alignment.bottomRight,
          colors: [Color(0xFFFFFFFF), Color(0xFFD9E2EC)]),
      clair: true),
  Couleur('Multicolore', degrade: _multicolore),
];

// Le wax / imprimé et le multicolore de la Maison — des dégradés à part.
const _waxImprime = SweepGradient(colors: [
  Color(0xFFF77F00),
  Color(0xFFFACC15),
  Color(0xFF16A34A),
  Color(0xFF2563EB),
  Color(0xFFDC2626),
  Color(0xFFF77F00),
]);
const _multiMaison = SweepGradient(colors: [
  Color(0xFFEC4899),
  Color(0xFF8B5CF6),
  Color(0xFF2563EB),
  Color(0xFF16A34A),
  Color(0xFFFACC15),
  Color(0xFFEC4899),
]);

/// La palette de base de la Maison — plus riche (crème, taupe, bleu nuit,
/// bordeaux, inox, wax) que les quinze générales.
const List<Couleur> paletteMaison = [
  Couleur('Blanc', couleur: Color(0xFFFFFFFF), clair: true),
  Couleur('Noir', couleur: Color(0xFF1B1A17)),
  Couleur('Gris', couleur: Color(0xFF9CA3AF)),
  Couleur('Beige', couleur: Color(0xFFE8D8C3), clair: true),
  Couleur('Crème', couleur: Color(0xFFF5EEE0), clair: true),
  Couleur('Marron', couleur: Color(0xFF7A4423)),
  Couleur('Taupe', couleur: Color(0xFF8B7D6B)),
  Couleur('Bleu', couleur: Color(0xFF2563EB)),
  Couleur('Bleu nuit', couleur: Color(0xFF1E3A5F)),
  Couleur('Vert', couleur: Color(0xFF16A34A)),
  Couleur('Rouge', couleur: Color(0xFFDC2626)),
  Couleur('Bordeaux', couleur: Color(0xFF7A1F2B)),
  Couleur('Orange', couleur: Color(0xFFF97316)),
  Couleur('Jaune', couleur: Color(0xFFFACC15), clair: true),
  Couleur('Rose', couleur: Color(0xFFEC4899)),
  Couleur('Violet', couleur: Color(0xFF8B5CF6)),
  Couleur('Doré', degrade: _dore, clair: true),
  Couleur('Argenté / inox', degrade: _argente, clair: true),
  Couleur('Wax / imprimé', degrade: _waxImprime),
  Couleur('Multicolore', degrade: _multiMaison),
];

/// Les teintes de bois — pour ce qui se vend AU BOIS et non à la couleur
/// (meubles, essences : teck, iroko, wengé…).
const List<Couleur> paletteBois = [
  Couleur('Bois naturel clair', couleur: Color(0xFFD6B98C), clair: true),
  Couleur('Bois miel', couleur: Color(0xFFB8823C)),
  Couleur('Teck', couleur: Color(0xFF9C6B32)),
  Couleur('Iroko', couleur: Color(0xFF8A5A2B)),
  Couleur('Acajou', couleur: Color(0xFF6B3A26)),
  Couleur('Wengé / brun foncé', couleur: Color(0xFF3E2A20)),
  Couleur('Ébène / noir', couleur: Color(0xFF1B1A17)),
  Couleur('Blanc laqué', couleur: Color(0xFFFFFFFF), clair: true),
  Couleur('Gris', couleur: Color(0xFF9CA3AF)),
  Couleur('Bicolore bois et blanc',
      degrade: LinearGradient(
          begin: Alignment.topLeft,
          end: Alignment.bottomRight,
          colors: [Color(0xFFB8823C), Color(0xFFFFFFFF)],
          stops: [0.5, 0.5]),
      clair: true),
];

// L'imprimé enfant — un dégradé pastel propre au rayon Bébé.
const _imprimeEnfant = SweepGradient(colors: [
  Color(0xFFF9A8D4),
  Color(0xFF93C5FD),
  Color(0xFFFACC15),
  Color(0xFFA7F3D0),
  Color(0xFFF9A8D4),
]);

/// La palette du rayon Bébé & Enfant — des pastels (rose layette, bleu ciel,
/// vert d'eau) et les imprimés enfant, que les teintes générales n'ont pas.
const List<Couleur> paletteBebe = [
  Couleur('Blanc', couleur: Color(0xFFFFFFFF), clair: true),
  Couleur('Écru / crème', couleur: Color(0xFFF5EEE0), clair: true),
  Couleur('Beige', couleur: Color(0xFFE8D8C3), clair: true),
  Couleur('Gris', couleur: Color(0xFF9CA3AF)),
  Couleur('Noir', couleur: Color(0xFF1B1A17)),
  Couleur('Rose', couleur: Color(0xFFF9A8D4), clair: true),
  Couleur('Rose vif', couleur: Color(0xFFEC4899)),
  Couleur('Bleu ciel', couleur: Color(0xFF93C5FD), clair: true),
  Couleur('Bleu', couleur: Color(0xFF2563EB)),
  Couleur('Bleu marine', couleur: Color(0xFF1E3A5F)),
  Couleur('Vert', couleur: Color(0xFF16A34A)),
  Couleur('Vert d’eau', couleur: Color(0xFFA7F3D0), clair: true),
  Couleur('Jaune', couleur: Color(0xFFFACC15), clair: true),
  Couleur('Orange', couleur: Color(0xFFF97316)),
  Couleur('Rouge', couleur: Color(0xFFDC2626)),
  Couleur('Violet', couleur: Color(0xFF8B5CF6)),
  Couleur('Marron', couleur: Color(0xFF7A4423)),
  Couleur('Wax / pagne', degrade: _waxImprime),
  Couleur('Imprimé enfant', degrade: _imprimeEnfant),
  Couleur('Multicolore', degrade: _multiMaison),
];

/// La clé d'attribut d'une variante : `var_<Couleur>_<champ>`.
String cleVariante(String nom, String champ) => 'var_${nom}_$champ';

/// « Noir, Bleu » → [Couleur, Couleur], dans l'ordre coché, en ne gardant que
/// les teintes présentes dans la palette active (les noms inconnus sont ignorés).
List<Couleur> lireCouleurs(dynamic v, List<Couleur> palette) {
  final noms = <String>[];
  if (v is List) {
    noms.addAll(v.map((e) => e.toString()));
  } else if (v is String) {
    noms.addAll(v.split(',').map((s) => s.trim()));
  }
  final out = <Couleur>[];
  for (final nom in noms) {
    if (nom.isEmpty) continue;
    for (final c in palette) {
      if (c.nom == nom) {
        out.add(c);
        break;
      }
    }
  }
  return out;
}
